course-material/interactive-queries-and-parameterization/introduction-to-ipywidgets.md

```diff
@@ -1,15 +1,376 @@
+# Loading in SQL extension
+%reload_ext sql
+# Initiating a DuckDB database named 'bank.duck.db' to run our SQL queries on
+%sql duckdb:///bank.duck.db
+```
+
+## Creating Table
+
+Let's start off with loading our `bank_cleaned.csv` file from our local directory to our newly created DuckDB database. Here we `CREATE OR REPLACE TABLE` in DuckDB called 'bank' `FROM` our `bank_cleaned.csv` file. The `read_csv_auto` is a function that helps SQL understand our local .csv file for creation into our database.
+
+```{code-cell} ipython3
+%%sql
+CREATE OR REPLACE TABLE bank AS
+FROM read_csv_auto('bank_cleaned.csv', header=True, sep=',')
+```
+
+## `%sql --interact {{widget_variable}}`
+
+First, you need to define the variable as the form of a basic data type or `ipywidgets` Widget.
+Then, pass the variable name into the `--interact` argument and use a `WHERE` clause to filter using the specified widgets variables. In this section, we shall delve into the different types of widgets and how to use them.
+
+### Basic Data Types
+
+The simplest way is to declare a variable with basic data types (Numeric, Text, Boolean...). [ipywidgets](https://ipywidgets.readthedocs.io/en/stable/examples/Using%20Interact.html?highlight=interact#Basic-interact) autogenerates UI controls for that variable.
+
+An example with a numeric variable `duration`, which creates a slider for its values as a UI, is as follows:
+
+```{code-cell} ipython3
+duration_min = 1500
+%sql --interact duration_min SELECT * FROM bank WHERE duration > {{duration_min}} LIMIT 5
+```
+
+<b>Note above</b> that we are filtering records by the variable `duration` if it is <b>greater than</b> the value of the slider widget `duration_min`.
+
+An example with a categorical variable `loan`, which creates a text box for the user to type in values as a UI, is as follows:
+
+```{code-cell} ipython3
+loan = "yes"  # Try inputting "no" in the output's text box
+%sql --interact loan SELECT * FROM bank WHERE loan == '{{loan}}' LIMIT 5
+```
+
+### Numeric Widgets
+
+There are many widgets distributed with `ipywidgets` that are designed to display numeric values. Widgets exist for displaying integers and floats, both bounded and unbounded. The integer widgets share a similar naming scheme to their floating point counterparts. By replacing "Float" with "Int" in the widget name, you can find the Integer equivalent.
+
+#### `IntSlider` and `FloatSlider`
+
+Numeric widgets provide further flexibility over basic data types. For example, the `IntSlider` and the `FloatSlider`, the simplest numeric widgets, can be employed to filter <b>integer</b> and <b>float</b> values respectively within a range (`min` and `max`) of your choice along with a `step` size and the `value` at which the slider is initialized. Moreover, we can create a bounded slider, indicated by the `<=` operator, or an unbounded slider, indicated by the `>=` operator, to filter the dataset within or outside our range of values respectively.
```

Comment:
Jonathan would have covered this already. Maybe add a "refresher" but no need to redefine it